### PR TITLE
ds1 editor

### LINKED ISSUE #79: Add editor features to DS1 viewer
Depends on:
* https://github.com/OpenDiablo2/OpenDiablo2/issues/909
* #68

### PR DESCRIPTION
**Pull request by gucio321**
fixes #79

### CODE REVIEW

File: hswidget/ds1_editor_widget.go [hunk 1]
```diff
@@ -0,0 +1,1004 @@
...
+			}),
+		})
+	}
+
+	return giu.Layout{
+		l,
+		giu.Separator(),
+		giu.Button("Add File##"+p.id+"AddFile").Size(actionButtonW, actionButtonH).OnClick(func() {
+			state.state = ds1EditorStateAddFile
+		}),
```
Comment (gucio321) [inline hswidget/ds1_editor_widget.go]: we can probably call some instance of mpq explorer

File: hswidget/ds1_editor_widget.go [hunk 2]
```diff
@@ -0,0 +1,1115 @@
...
+			layerStream[layerIdx] = d2enum.LayerStreamShadow
+			layerIdx++
+		}
+		if p.ds1.NumberOfSubstitutionLayers > 0 {
+			layerStream[layerIdx] = d2enum.LayerStreamSubstitute
+		}
+	}
+
+	p.ds1.LayerStreamTypes = layerStream
+}
```
Comment (gucio321) [inline hswidget/ds1_editor_widget.go]: We'll be able to remove after merging https://github.com/OpenDiablo2/OpenDiablo2/pull/1059

File: hswidget/ds1widget/helper.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// Warning: this is 1:1 copy from
+// github.com/OpenDiablo2/OpenDiablo2/d2common/d2fileformats/d2ds1.(*DS1).setupLayerStreamType()
+// but this method is unexported for now, so...
+// see https://github.com/OpenDiablo2/OpenDiablo2/pull/1059
+func (p *DS1Widget) recreateLayerStreamTypes() {
+	var layerStream []d2enum.LayerStreamType
+
+	// nolint:gomnd // this is constant version
+	// see in OpenDiablo2
+	if p.ds1.Version < 4 {
```
Comment (gravestench) [inline hswidget/ds1widget/helper.go:16]: #1059 is merged

File: hswidget/ds1widget/state.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+// DS1AddObjectState represents state of new object
+type DS1AddObjectState struct {
+	objType  int32
+	objID    int32
+	objX     int32
+	objY     int32
+	objFlags int32
+}
```
Comment (gravestench) [inline hswidget/ds1widget/state.go]: is there any reason for these types to be exported?

File: hswidget/ds1widget/state.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+// DS1State represents ds1 viewers state
+type DS1State struct {
+	*ds1Controls
+	mode                ds1EditorMode
+	confirmDialog       *hswidget.PopUpConfirmDialog
+	newFilePath         string
+	addObjectState      DS1AddObjectState
+	addPathState        DS1AddPathState
+	addFloorShadowState DS1AddFloorShadowState
+	addWallState        DS1AddWallState
```
Comment (gravestench) [inline hswidget/ds1widget/state.go]: yeah, those types probably shouldn't be exported

File: hswidget/ds1widget/tile_records.go [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+	for y := range p.ds1.Tiles {
+		for x := range p.ds1.Tiles[y] {
+			newWalls := make([]d2ds1.WallRecord, 0)
+
+			for n, i := range p.ds1.Tiles[y][x].Walls {
+				if n != int(state.object) {
+					newWalls = append(newWalls, i)
+				}
+			}
```
Comment (gravestench) [inline hswidget/ds1widget/tile_records.go]: probably should change `n` and `i` to something more descriptive

File: hswidget/ds1widget/tile_records.go [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+	state.addFloorShadowState.cb = func() {
+		newShadow := p.createFloorShadowRecord()
+
+		for y := range p.ds1.Tiles {
+			for x := range p.ds1.Tiles[y] {
+				p.ds1.Tiles[y][x].Shadows = make([]d2ds1.FloorShadowRecord, 1)
+				p.ds1.Tiles[y][x].Shadows[0] = newShadow
+			}
+		}
+	}
```
Comment (gravestench) [inline hswidget/ds1widget/tile_records.go]: is this setting the `Shadows` for all tiles to the `newShadow` record when you add a new `shadow`... ?

Comment (gravestench) [reply hswidget/ds1widget/tile_records.go]: what if the tile already had a shadow...?

File: hswidget/ds1widget/helper.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"github.com/OpenDiablo2/OpenDiablo2/d2common/d2enum"
+)
+
+// Warning: this is 1:1 copy from
+// github.com/OpenDiablo2/OpenDiablo2/d2common/d2fileformats/d2ds1.(*DS1).setupLayerStreamType()
+// but this method is unexported for now, so...
+// see https://github.com/OpenDiablo2/OpenDiablo2/pull/1059
+func (p *DS1Widget) recreateLayerStreamTypes() {
```
Comment (gravestench) [inline hswidget/ds1widget/helper.go]: 1059 has been closed; this is a bad idea to put state management code for `d2ds1` outside of `d2ds1`, and an even worse idea to make users of the package responsible for knowing when to call state management/upkeep methods.

we need to take a closer look at `d2ds1` and resolve state management there before we merge this PR.

**Review by gravestench — CHANGES_REQUESTED** [review]
We need to fix our state management issues in `d2ds1` before we merge this PR. It should not be a responsibility of the user of `d2ds1` to call state management methods when they edit fields in the ds1 struct. State management should be internal to `d2ds1`

File: hswidget/ds1widget/state.go (hunk 4 above)
Comment (gucio321) [reply hswidget/ds1widget/state.go]: any one, forgot to change

File: hswidget/ds1widget/state.go (hunk 5 above)
Comment (gucio321) [reply hswidget/ds1widget/state.go]: absolutly

File: hswidget/ds1widget/tile_records.go (hunk 6 above)
Comment (gucio321) [reply hswidget/ds1widget/tile_records.go]: well, n (natural number) is ok, but `i` could be replaced

Comment (gravestench) [reply hswidget/ds1widget/tile_records.go]: yeah, i'm okay with that. I probably wrote that code to begin with :x

File: hswidget/ds1widget/tile_records.go (hunk 7 above)
Comment (gucio321) [reply hswidget/ds1widget/tile_records.go]: - when one tile has a shadow, then another tiles too (`DS1.LayerStreamTypes`)
- it is only called, when tile hasn't any shadow

Comment (gravestench) [reply hswidget/ds1widget/tile_records.go]: Ok, after looking at the tile record struct, i think i get it; tiles have a variable number of walls, floors, and shadows.

**Review by gravestench — APPROVED** [review]
Good work!

**Pull request merged**